File: lab_2_C#/MinFinderTask.cs
class MinFinderTask
{
    private readonly int[] array;
    private readonly int start;
    private readonly int end;
    private readonly ParallelMinFinder finder;

    public MinFinderTask(int[] array, int start, int end, ParallelMinFinder finder)
    {
        this.array = array;
        this.start = start;
        this.end = end;
        this.finder = finder;
    }

    public void Run()
    {
        int localMin = int.MaxValue;
        int localMinIndex = -1;

        for (int i = start; i < end; i++)
        {
            if (array[i] < localMin)
            {
                localMin = array[i];
                localMinIndex = i;
            }
        }

        finder.ReportResult(localMin, localMinIndex);
    }
}

File: lab_2/src/MinFinderTask.java
public class MinFinderTask implements Runnable {
    private final int[] array;
    private final int start;
    private final int end;
    private final SharedState state;
    private final int totalThreads;

    public MinFinderTask(int[] array, int start, int end, SharedState state, int totalThreads) {
        this.array = array;
        this.start = start;
        this.end = end;
        this.state = state;
        this.totalThreads = totalThreads;
    }

    @Override
    public void run() {
        int localMin = Integer.MAX_VALUE;
        int localMinIndex = -1;

        for (int i = start; i < end; i++) {
            if (array[i] < localMin) {
                localMin = array[i];
                localMinIndex = i;
            }
        }

        synchronized (state.getLock()) {
            state.updateMin(localMin, localMinIndex);
            state.incrementCompletedThreads();

            if (state.getCompletedThreads() == totalThreads) {
                state.getLock().notify();
            }
        }
    }
}
